### PR TITLE
WIP: Enable DateTime completion in string interpolation format strings.

### LINKED ISSUE #43267: DateTime completion should work in string interpolation format strings.
i.e. here: `$"Hello, {name}! Today is {date.DayOfWeek}, it's {date:HH:mm} now."`

Should be fairly straightforward to do.  Just need to update the code to not just expect string literals, but also interpolation format strings where the expr type is the right form.

### PR DESCRIPTION
**Pull request by zaytsev-victor**
Fixes #43267

### CODE REVIEW

File: src/Features/Core/Portable/EmbeddedLanguages/DateAndTime/DateAndTimeEmbeddedCompletionProvider.cs [hunk 1]
```diff
@@ -76,8 +76,8 @@ public override async Task ProvideCompletionsAsync(CompletionContext context)
                 return;
 
             var stringToken = stringTokenOpt.Value;
-            if (position <= stringToken.SpanStart || position >= stringToken.Span.End)
-                return;
+            //if (position <= stringToken.SpanStart || position >= stringToken.Span.End)
+            //    return;
```
Comment (zaytsev-victor) [inline src/Features/Core/Portable/EmbeddedLanguages/DateAndTime/DateAndTimeEmbeddedCompletionProvider.cs:80]: I'm not sure what to do with this line. CompletionProvider receives position of '}' of format expression inside of string interpolation. I get previous token for analization, but SpanStart for the token is ```position - 1```. I've temporarily commented this line. @CyrusNajmabadi  Do you have any idea what the best way to handle it?

File: src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs [hunk 2]
```diff
@@ -92,13 +98,7 @@ internal sealed class DateAndTimePatternDetector
             if (!IsMethodArgument(token, syntaxFacts))
                 return false;
 
-            if (!syntaxFacts.IsLiteralExpression(token.Parent))
-                return false;
-
-            if (!syntaxFacts.IsArgument(token.Parent.Parent))
-                return false;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs:99]: why were these removed.

File: src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs [hunk 3]
```diff
@@ -133,32 +137,41 @@ private static bool IsMethodArgument(SyntaxToken token, ISyntaxFacts syntaxFacts
...
-            if (TryAnalyzeInvocation(method, argName, argIndex))
-                return true;
-
-            foreach (var candidate in symbolInfo.CandidateSymbols)
+            if (IsPossiblyDateAndTimeTokenInStringInterpolation(token, _info.SyntaxFacts))
             {
-                if (TryAnalyzeInvocation(candidate, argName, argIndex))
-                    return true;
+                var type = _semanticModel.GetTypeInfo(
+                    _info.SyntaxFacts.GetExpressionOfInterpolation(token.Parent!.Parent),
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs:171]: could IsPossiblyDateAndTimeTokenInStringInterpolation return teh values you need so you don't need to `!` things.

File: src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs [hunk 4]
```diff
@@ -187,7 +200,7 @@ private bool TryAnalyzeInvocation(ISymbol? symbol, string? argName, int? argInde
                IsDateTimeType(method.ContainingType) &&
                AnalyzeStringLiteral(method, argName, argIndex);
 
-        private bool IsDateTimeType(INamedTypeSymbol containingType)
+        private bool IsDateTimeType(ITypeSymbol containingType)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs:203]: why?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/EmbeddedLanguages/VirtualChars/CSharpVirtualCharService.cs [hunk 5]
```diff
@@ -58,15 +58,19 @@ protected override VirtualCharSequence TryConvertToVirtualCharsWorker(SyntaxToke
...
+                    // The sections between  `}` and `{` are InterpolatedStringTextToken *as are* the
+                    // format specifiers in an interpolated string.  We only want to get the virtual
+                    // chars for this first type.
                     return interpolatedString.StringStartToken.Kind() == SyntaxKind.InterpolatedVerbatimStringStartToken
                        ? TryConvertVerbatimStringToVirtualChars(token, "", "", escapeBraces: true)
                        : TryConvertStringToVirtualChars(token, "", "", escapeBraces: true);
                 }
+                else if (token.Parent is InterpolationFormatClauseSyntax)
+                {
+                    return TryConvertStringToVirtualChars(token, "", "", escapeBraces: true);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/EmbeddedLanguages/VirtualChars/CSharpVirtualCharService.cs:72]: are these the actual semantics we want here? i.e. what syntax is allowed inside the format clause.  is it the same as in the rest of the interpolated string?  what if it's a verbatim interpolated string?

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/EmbeddedLanguages/VirtualChars/CSharpVirtualCharService.cs:72]: based on the actual impl, the above is not correct:

```
        private SyntaxToken MakeStringToken(string text, string bodyText, bool isVerbatim, SyntaxKind kind)
        {
            var prefix = isVerbatim ? "@\"" : "\"";
            var fakeString = prefix + bodyText + "\"";
            using (var tempLexer = new Lexer(Text.SourceText.From(fakeString), this.Options, allowPreprocessorDirectives: false))
            {
                LexerMode mode = LexerMode.Syntax;
                SyntaxToken token = tempLexer.Lex(ref mode);
```

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/EmbeddedLanguages/VirtualChars/CSharpVirtualCharService.cs:72]: the format clause follows the same rules wrt verbatim-nes as the rest of the interpolated string.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/EmbeddedLanguages/VirtualChars/CSharpVirtualCharService.cs [hunk 6]
```diff
@@ -58,15 +58,19 @@ protected override VirtualCharSequence TryConvertToVirtualCharsWorker(SyntaxToke
...
+                }
             }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/EmbeddedLanguages/VirtualChars/CSharpVirtualCharService.cs:74]: no equivalent change for VB?

File: src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs (hunk 2 above)
Comment (zaytsev-victor) [reply src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs:99]: ```IsMethodArgument``` contains these conditions:
```csharp
private static bool IsMethodArgument(SyntaxToken token, ISyntaxFacts syntaxFacts)
            => syntaxFacts.IsLiteralExpression(token.Parent) &&
               syntaxFacts.IsArgument(token.Parent!.Parent);
```

File: src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs [hunk 7]
```diff
@@ -133,32 +137,41 @@ private static bool IsMethodArgument(SyntaxToken token, ISyntaxFacts syntaxFacts
 
         public bool IsDateAndTimeToken(SyntaxToken token, CancellationToken cancellationToken)
         {
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs:139]: this turned out to be super confusing.  why not just have the core method be: IsDateAndTimeToken(...) => IsDateAndTimeTokenInInvocation(...) || IsDateAndTimeTokenInFormatClause(...)`.  Then pull all the existing logic into the new IsDateAndTimeTokenInInvocation (unchanged otherwise), and keep the other check independent?

File: src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs [hunk 8]
```diff
@@ -116,6 +116,10 @@ internal sealed class DateAndTimePatternDetector
             return true;
         }
 
+        private static bool IsPossiblyDateAndTimeTokenInStringInterpolation(
+            SyntaxToken token, ISyntaxFacts syntaxFacts)
+            => syntaxFacts.IsInterpolationFormatSyntax(token.Parent);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs:121]:
```suggestion
        private static bool IsPossiblyDateAndTimeTokenInStringInterpolation(SyntaxToken token, ISyntaxFacts syntaxFacts)
            => syntaxFacts.IsInterpolationFormatSyntax(token.Parent);
```

File: src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs [hunk 9]
```diff
@@ -92,13 +98,7 @@ internal sealed class DateAndTimePatternDetector
...
                 return false;
 
-            if (!syntaxFacts.IsLiteralExpression(token.Parent))
-                return false;
-
-            if (!syntaxFacts.IsArgument(token.Parent.Parent))
-                return false;
-
-            var argumentList = token.Parent.Parent.Parent;
+            var argumentList = token.Parent!.Parent!.Parent;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs:101]: i don't like the `!`s at all.  the benefit of hte previous system was that we didn't need this.

File: src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs [hunk 10]
```diff
@@ -74,10 +74,16 @@ internal sealed class DateAndTimePatternDetector
 
         /// <summary>
         /// Checks if this <paramref name="token"/> is possibly a string literal token that could contain a date or time
-        /// format string passed into an method call.  If so, <paramref name="argumentNode"/> and <paramref
-        /// name="invocationExpression"/> will be the argument and invocatoin the string literal was passed as.
+        /// format string passed.
         /// </summary>
         public static bool IsPossiblyDateAndTimeToken(
+            SyntaxToken token, ISyntaxFacts syntaxFacts)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs:80]:
```suggestion
        public static bool IsPossiblyDateAndTimeToken(SyntaxToken token, ISyntaxFacts syntaxFacts)
```

File: src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimeEmbeddedLanguage.cs [hunk 11]
```diff
@@ -29,9 +29,9 @@ public DateAndTimeEmbeddedLanguage(EmbeddedLanguageInfo info)
             Document document, int position, CancellationToken cancellationToken)
         {
             var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
-            var token = root.FindToken(position);
+            var token = root.FindTokenOnLeftOfPosition(position);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimeEmbeddedLanguage.cs:32]: why?

File: src/Features/Core/Portable/EmbeddedLanguages/DateAndTime/DateAndTimeEmbeddedCompletionProvider.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/EmbeddedLanguages/DateAndTime/DateAndTimeEmbeddedCompletionProvider.cs:80]: i'm not really understanding the issue.  stringTokenOpt should be the appropriate string token for either case, unless the caller isn't passing the right thing.

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests_DateAndTime.vb [hunk 12]
```diff
@@ -121,5 +121,120 @@ class c
...
+{
+    void goo(DateTime d)
+    {
+        _ = $"Text {date:f$$}";
+    }
+}
+]]></Document>, showCompletionInArgumentLists:=showCompletionInArgumentLists)
+
+                state.SendTypeChars("f")
+                Await state.AssertNoCompletionSession()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests_DateAndTime.vb:236]: VB tests?

File: src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs:139]: note that that's what you did for IsPossiblyDateAndTimeToken which was very clear.

File: src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs (hunk 9 above)
Comment (zaytsev-victor) [reply src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs:101]: Then it is better to remove ```IsMethodArgument```, right?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimePatternDetector.cs:101]: i woudl be fine with that.

File: src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimeEmbeddedLanguage.cs (hunk 11 above)
Comment (zaytsev-victor) [reply src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimeEmbeddedLanguage.cs:32]: To get ```InterpolatedStringTextToken```. For ```ToString``` method it returns the same token.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimeEmbeddedLanguage.cs:32]: Seems weird. If I was after the token in a format clause, what token is returned?

Comment (zaytsev-victor) [reply src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimeEmbeddedLanguage.cs:32]: ```root.FindToken(position)``` returns ```}``` for ```$"Text {date:$$}"``` and for ```$"Text {date:M$$}"```

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimeEmbeddedLanguage.cs:32]: ok.  but what does it return for FindTokenOnLeftOfPosition?  i would expect it to return the `:` token then.  Is that not hte case?  If FindToken returns the `}` chec if that's the case (i.e. you're at the start of that token, and move back one token.  That seems like the simpler approach to take.

Comment (zaytsev-victor) [reply src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimeEmbeddedLanguage.cs:32]: ```FindTokenOnLeftOfPosition``` returns ```:``` for ```$"Text {date:$$}"``` and ```:M``` for ```$"Text {date:$$}"```. That exactly what we need here.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimeEmbeddedLanguage.cs:32]: ok.  can you please doc in the code.  thanks :)

**Comment by CyrusNajmabadi** [comment]
@zaytsev-victor What woudl you like to do with this PR?  Thanks! :)

**Comment by zaytsev-victor** [comment]
@CyrusNajmabadi I will be very busy until the end of this week but I can finish it on the next week.
 
I need to add comment for line [var token = root.FindTokenOnLeftOfPosition(position);](https://github.com/dotnet/roslyn/blob/e35a0c18fc0fa970c42bcc986543bc6a7b973eaf/src/Workspaces/Core/Portable/EmbeddedLanguages/DateAndTime/LanguageServices/DateAndTimeEmbeddedLanguage.cs#L32).

Could you please help me with this [line](https://github.com/dotnet/roslyn/blob/6c29c0fb0dc64cfbe24f70d0a30290d87a246186/src/Features/Core/Portable/EmbeddedLanguages/DateAndTime/DateAndTimeEmbeddedCompletionProvider.cs#L79-80)?

This isn't work in string interpolation case. Should I delete it or rework somehow?

**Comment by CyrusNajmabadi** [comment]
> Could you please help me with this line?

**Comment by CyrusNajmabadi** [comment]
Hey @zaytsev-victor will you be able to get to this?  or would you want us to take this over for you?  Thanks!

**Comment by CyrusNajmabadi** [comment]
Hey @zaytsev-victor will you be able to get to this? or would you want us to take this over for you? Thanks!

**Comment by CyrusNajmabadi** [comment]
I'm going to look at thsi to see what it woudl take to support this scenario.

**Comment by CyrusNajmabadi** [comment]
@zaytsev-victor thanks for the PR!
